### PR TITLE
Week-12 Homework

### PR DESCRIPTION
**Pull request by TaiChiLe**

### CODE REVIEW

File: week-12/homework/homework-1.html [hunk 1]
```diff
@@ -0,0 +1,229 @@
+<html lang="en">
+
```
Comment (jack-vo) [inline week-12/homework/homework-1.html:2]: @TaiChiLe I do not think your VS Auto format on save is working correctly as the there is a lot of empty line and does not look like the page is formatted correctly

File: week-12/homework/index.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
+let input = document.querySelector('.input');
```
Comment (jack-vo) [inline week-12/homework/index.js:1]: Same for this JS file, the file formatting does not look right to me, let's sharing next time so I can fix up your Prettier

File: week-12/homework/index.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+let addBtn = document.querySelector('.add-btn');
+let itemContainer = document.querySelector('.item-container');
+let deleteBtns = document.querySelectorAll('.delete-btn');
+let editBtns = document.querySelectorAll('.edit-btn');
+let currentTaskNum = 0;
+
+//Creates tasklist, takes taskNumber and input value
+let processTaskList = function (taskNum, inputValue) {
+    let newInput = document.createElement("input");
```
Comment (jack-vo) [inline week-12/homework/index.js]: This whole process below can be optimized to make them a lot less "manual" and easier to maintain in the future.
https://github.com/TaiChiLe/fullstack-learning/pull/2/files#diff-2bbe4e80b92dad33be1498200f3201df02ec5be47bcfe1a0840263cc5c12bd55R10-R48

Check my recommendation below using `innerHTML` and `string`
```
itemContainer.innerHTML =  itemContainer.innerHTML + `
<div class="item d-flex pt-1" item-number="${taskNum}">
    <input class="input-item form-control" type="text" disabled input-item-number="${taskNum}"/>
    <button class="delete-btn btn btn-danger" delte-item-number="${taskNum}">X</button>
    ... you work out the remaining elements yourself
</div>
```

This way, it's a lot faster for you to put together your entire element list and minize the chance of wrong code

File: week-12/homework/index.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    let deleteBtn = document.createElement("button");
```
Comment (jack-vo) [inline week-12/homework/index.js]: For simple string, always use `'`, do not use `"`

File: week-12/homework/index.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+    newDiv.appendChild(newInput);
+    newDiv.appendChild(deleteBtn);
+    newDiv.appendChild(editBtn);
+    newDiv.appendChild(saveBtn);
+
+    itemContainer.appendChild(newDiv);
+
+
+    input.value = '';
```
Comment (jack-vo) [inline week-12/homework/index.js:16]: Remember that, `.querySelector...` exists on all HTML elements. for example `newDiv.querySelector` ...

Depending on which HTML element you call, the selection will happen to children elements belonged to that element.

For example, `newDiv.querySelect('.my-button')` means get the first `.my-button` element that is a child of `newDiv`

Once you have attach the element using innerHTML, you can start doing element selection below
```
let input = itemContainer.querySelector('input');

input.value = '';
```

File: week-12/homework/index.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+    deleteBtns = document.querySelectorAll('.delete-btn');
```
Comment (jack-vo) [inline week-12/homework/index.js]: The logic below are not correct as each time you call `processTaskList`, you add MORE click event to all the delete buttons. 
What does it mean? `addEventListener` will add another new `event` (in this case `"click"`) to the element.

```
myButton.addEventListener('click', firstClick);
myButton.addEventListener('click', secondClick);
```

when you actually `click` the `myButton`, 2 functions will be executed -> `firstClick` then `secondClick`

So with the code below the `processsDeleteBtn` will be called many times depending on how many times you click the `Add` button

File: week-12/homework/index.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    editBtns = document.querySelectorAll('.edit-btn');
+
+    for (let i = 0; i < deleteBtns.length; i++) {
```
Comment (jack-vo) [inline week-12/homework/index.js]: There is a much better way so you do not need to loop and at the same time ensure we attach the right event only once.

I list my suggestion below and you need to think about how you will execute it
- 1. When this function is execute, create an empty `div`
- 2. Use the approach `innerHTML` above to set HTML content you need to that `div`. PS. if you follow this way you do not need any data-attribute set to input or button or anything
- 3. Append the `div` above to `itemContainer`
- 4. Start selecting buttons and input inside that `div` using `newDiv.querySelector`
- 5. Step 4 allow us to select just the buttons and input inside that `newDiv` so you just need to call `deleteBtn.addEventListener('click', processDeleteBtn)` ... 
- Work out the remaining logic yourself

**Review by jack-vo — CHANGES_REQUESTED** [review]
Nice try 👍 
The TODO list JS can be improved further so check my comments and give it another go, @TaiChiLe

File: week-12/homework/homework-1.html (hunk 1 above)
Comment (TaiChiLe) [reply week-12/homework/homework-1.html:2]: I think i put in these empty lines

**Pull request merged**